make_count/make_count.tex

```diff
@@ -0,0 +1,293 @@
+
+\setlength{\parindent}{0.0in}
+\setlength{\parskip}{0.05in}
+\title{Privacy Proofs for OpenDP: MakeCount}
+\author{Connor Wagaman -- wagaman@college.harvard.edu}
+
+\usepackage{xcolor}
+\usepackage{amsthm}
+\newtheorem{theorem}{Theorem}[section]
+\newtheorem{lemma}[theorem]{Lemma}
+
+\theoremstyle{definition}
+\newtheorem{remark}{Remark}
+\newtheorem{definition}{Definition}[section]
+\newtheorem{observation}{Observation}
+\newtheorem{note}{Note}
+\newtheorem{hope}{Hope}
+\newtheorem{warning}{Warning}
+\newtheorem{problem}{Problem}
+\newtheorem{fear}{Fear}
+
+\begin{document}
+
+\maketitle
+
+\tableofcontents
+
+\section{MakeCount}
+
+\subsection{Implementation of MakeCount in Rust}
+\label{sub:rust-implementation}
+
+\inOpenDPRust{make\_count}{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/count.rs}
+
+\permalink{https://github.com/opendp/opendp/blob/c3b5c3bd9fc50c556362b628f08c5fddea069b4d/rust/opendp/src/trans/count.rs\#L14-L27}{12/07/2021} (It is from \href{https://github.com/opendp/opendp/pull/173}{this pull request}.)
+The Rust code can also be seen below. 
+
+\begin{lstlisting}[language = Rust]
+pub fn make_count<TIA, TO>(
+) -> Fallible<Transformation<VectorDomain<AllDomain<TIA>>, AllDomain<TO>, SymmetricDistance, AbsoluteDistance<TO>>>
+    
+    where TO: ExactIntCast<usize> + One + DistanceConstant<IntDistance>, IntDistance: InfCast<TO> {
+    
+    Ok(Transformation::new(
+        VectorDomain::new_all(),
+        AllDomain::new(),
+        // think of this as: min(arg.len(), TO::MAX_CONSECUTIVE)
+        Function::new(move |arg: &Vec<TIA>|
+            TO::exact_int_cast(arg.len()).unwrap_or(TO::MAX_CONSECUTIVE)),
+        SymmetricDistance::default(),
+        AbsoluteDistance::default(),
+        StabilityRelation::new_from_constant(TO::one())))
+}
+\end{lstlisting}
+
+\subsection{Implementation of MakeCount in Python-style pseudocode, with preconditions}
+\label{sec:python-pseudocode}
+
+We now use Python-style pseudocode to present a representation of the Rust function.
+
+\emph{Recall that functions in the pseudocode are defined in the document \href{https://www.overleaf.com/project/60d215bf90b337ac02200a99}{``List of definitions used in the pseudocode"}.}
+
+\emph{The use of \texttt{code}-style parameters in the preconditions section below (for example, \texttt{input\_domain}) means that this information should be passed along to the \texttt{Transformation} constructor.}
+
+Here, we use preconditions to check for traits, and to specify the domains and metrics.
+
+\textbf{Preconditions}
+\begin{itemize}
+    \item \textbf{User-specified types:} The \texttt{make\_count} function takes two inputs: a generic input type \texttt{TIA} for the \texttt{Transformation} (meaning that the input vector to \texttt{Transformation} is of type \texttt{Vec(TIA)}), and a generic output type \texttt{TO} for the \texttt{Transformation}.

```

Comment:
Do we mean to say that the input vector is from *data domain* Vec(TIA)?  It will be of type Vec(TIA)::Carrier, right?

Reply:
Salil's intention is right- that the input vector is a member of the data domain, and its type is the data domain's carrier. But the syntax is a little different:

The data domain is `VectorDomain<AllDomain<TIA>>`.

The following holds:
```VectorDomain<AllDomain<TIA>>::Carrier == Vec<TIA>```

I think the recommended action here is to constrain the input vector by the domain, instead of by the type.

Reply:
My impression from the comments above is that the current latex gives us what we need, so I am going to leave it as-is for now. Please let me know if I should, in fact, make changes.